### PR TITLE
Customize your Store > Implement retry mechanism, enhance prompts and fix issue with AI content not being generated for new products

### PR DESCRIPTION
**Pull request by nefeline**
## What

On this PR, a series of changes are made to ensure that:

- Whenever AI fails to generate content for Patterns, Products, or while managing images, a retry mechanism is triggered up to five times, with an error being triggered only after the fifth failed attempt.
- Prompts to AI are enhanced for products, patterns, and images.
- The method responsible for fetching the collection of dummy products to be updated by AI `fetch_dummy_products_to_update` is not invoked on the `product` endpoint but on the `products` endpoint instead. This solves the regression where this method would be invoked multiple times (for each product), resulting in too many products being programmatically updated on each request and not all of them having AI-managed/generated content assigned to them.
- Ensure the action for updating the business description is scheduled only if the provided description differs from the previously saved one.

The following diagrams demonstrate the updated flow with the changes made on this PR in place:

### AI images management:

image_flow_2

### AI content generation for products and patterns:

content_flow

## Why

These changes ensure AI content generation and image management are more consistent and reliable.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Setup

Please test the feature on a WooExpress install instead of Jurassic Ninja, as the latter can be very limited in terms of server resources and unstable, as other installs can interfere with your own.

- If you already have a pre-configured install, remove the pre-existing WooCommerce Blocks Plugin and upload the one from this zip file: [woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415580/woocommerce-gutenberg-products-block.zip)
- Alternatively, create a new WooExpress install from scratch 
- Remove the pre-existing WooCommerce plugin via SSH and
- Install and activate WooCommerce from the following zip file: 
[woocommerce.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415324/woocommerce.zip)
- Install and activate WooCommerce Blocks from the following zip file: [woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415580/woocommerce-gutenberg-products-block.zip)
- Install and activate WooCommerce Beta Tester (the plugin is available [within the monorepo](https://github.com/woocommerce/woocommerce/tree/trunk/plugins))
- Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag:

Screenshot 2023-10-22 at 10 32 24

### Test the CYS experience

0. Make sure you don't have any products in your store.
1. Head over to Home > Customize your Store:
 
Screenshot 2023-10-22 at 09 29 08

2. Click on design with AI.
3. Type a description for your business.
4. Make sure six new dummy products are created for your store.
5. Make sure the AI content is properly generated and images are assigned to patterns and products based on the description provided.

Caveat: the AI endpoint is known for being unstable, so you might encounter errors during the loading screen. If that happens, click "Try Again" to attempt updating one more time. We have additional upcoming changes to ensure better error handling planned for implementation in upcoming PRs.

* [x] Do not include in the Testing Notes 
* [x] Should be tested by the development team exclusively 

## Screenshots or screencast

https://github.com/woocommerce/woocommerce-blocks/assets/15730971/2ccecc54-d1a5-4517-b190-9697ff02a6c6

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:

* [ ] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

### CODE REVIEW

File: src/Patterns/PatternsHelper.php [hunk 1]
```diff
@@ -122,41 +123,63 @@ public static function upsert_patterns_ai_data_post( $patterns_dictionary ) {
...
 		}
 
-		$patterns_dictionary_file = plugin_dir_path( __FILE__ ) . 'dictionary.json';
-
-		if ( ! file_exists( $patterns_dictionary_file ) ) {
-			return new WP_Error( 'missing_patterns_dictionary', __( 'The patterns dictionary is missing.', 'woo-gutenberg-products-block' ) );
+		if ( $patterns_dictionary === $default_patterns_dictionary || empty( $patterns_dictionary ) ) {
+			if ( $pattern_slug ) {
+				return self::find_pattern_by_slug( $default_patterns_dictionary, $pattern_slug );
+			}
```
Comment (gigitux) [inline src/Patterns/PatternsHelper.php]: Could we remove the nested if in this way? 
```suggestion
		if ( ($patterns_dictionary === $default_patterns_dictionary || empty( $patterns_dictionary )) && $pattern_slug ) {
				return self::find_pattern_by_slug( $default_patterns_dictionary, $pattern_slug );
			}
```

File: src/Images/Pexels.php [hunk 2]
```diff
@@ -26,13 +27,73 @@ class Pexels {
 	 * @return array|\WP_Error Array of images, or WP_Error if the request failed.
 	 */
 	public function get_images( $ai_connection, $token, $business_description ) {
-		$search_term = $this->define_search_term( $ai_connection, $token, $business_description );
+		if ( str_word_count( $business_description ) === 1 ) {
+			$search_term = $business_description;
```
Comment (gigitux) [inline src/Images/Pexels.php:31]: It is not a strong opinion, but I would remove this check, given that it is a very rare use case, I would keep the logic simple.

File: src/Images/Pexels.php [hunk 3]
```diff
@@ -47,17 +108,72 @@ public function get_images( $ai_connection, $token, $business_description ) {
...
+
+		if ( ! is_array( $filtered_image_titles ) ) {
+			return new \WP_Error( 'ai_service_unavailable', __( 'AI Service is unavailable, try again later.', 'woo-gutenberg-products-block' ) );
+		}
+
+		foreach ( $returned_images as $returned_image ) {
+			if ( isset( $returned_image['title'] ) && ! in_array( $returned_image['title'], $filtered_image_titles, true ) ) {
+				unset( $returned_image );
+			}
+		}
```
Comment (gigitux) [inline src/Images/Pexels.php:173]: I would add a comment to explain which kind of check we are doing.

File: src/Patterns/PatternsHelper.php [hunk 4]
```diff
@@ -74,10 +74,11 @@ public static function get_image_url( array $images, int $index, string $default
 
 		return $image;
 	}
+
 	/**
 	 * Returns the post that has the generated data by the AI for the patterns.
 	 *
-	 * @return WP_Post|null
+	 * @return int|\WP_Post|null
```
Comment (gigitux) [inline src/Patterns/PatternsHelper.php]: Is there a chance that $post[0] is an int?

File: src/Patterns/PatternsHelper.php [hunk 5]
```diff
@@ -122,41 +123,63 @@ public static function upsert_patterns_ai_data_post( $patterns_dictionary ) {
...
+			if ( json_last_error() !== JSON_ERROR_NONE ) {
+				return new WP_Error( 'json_decode_error', __( 'Error decoding JSON.', 'woo-gutenberg-products-block' ) );
 			}
 		}
 
-		$patterns_dictionary_file = plugin_dir_path( __FILE__ ) . 'dictionary.json';
-
-		if ( ! file_exists( $patterns_dictionary_file ) ) {
-			return new WP_Error( 'missing_patterns_dictionary', __( 'The patterns dictionary is missing.', 'woo-gutenberg-products-block' ) );
+		if ( $patterns_dictionary === $default_patterns_dictionary || empty( $patterns_dictionary ) ) {
```
Comment (gigitux) [inline src/Patterns/PatternsHelper.php]: I would add some comments that explain which kind of check we are doing.

**Review by gigitux — APPROVED** [review]
Thanks for your brilliant and precious work! It's great! I'm approving the PR. I added some comments that it could be interesting to address to improve the clarity of the code/checks that we are doing, but if you think that it is not necessary, feel free to merge as it is!

File: src/Images/Pexels.php (hunk 2 above)
Comment (nefeline) [reply src/Images/Pexels.php:31]: I added it to save us some money during development: I've been testing with a single word so we can make one less API call to AI for defining the search term :)

File: src/Patterns/PatternsHelper.php (hunk 1 above)
Comment (nefeline) [reply src/Patterns/PatternsHelper.php]: Good call! Updated on https://github.com/woocommerce/woocommerce-blocks/pull/11759/commits/0ba01c8e7850d49364c3e84850319fbbd2305b84

File: src/Images/Pexels.php (hunk 3 above)
Comment (nefeline) [reply src/Images/Pexels.php:173]: Done! https://github.com/woocommerce/woocommerce-blocks/pull/11759/commits/0ba01c8e7850d49364c3e84850319fbbd2305b84

File: src/Patterns/PatternsHelper.php (hunk 4 above)
Comment (nefeline) [reply src/Patterns/PatternsHelper.php]: Nice catch, there isn't! Updated on https://github.com/woocommerce/woocommerce-blocks/pull/11759/commits/0ba01c8e7850d49364c3e84850319fbbd2305b84

File: src/Patterns/PatternsHelper.php (hunk 5 above)
Comment (nefeline) [reply src/Patterns/PatternsHelper.php]: In this case, IMO explaining these conditionals would be a bit redundant, so I kept it as-is.

**Comment by nefeline** [comment]
Thank you so much for your great review @gigitux 🎉 ! I've added comments & also addressed the feedback given.
